src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -4038,84 +4038,70 @@ - (MTRBaseDevice *)newBaseDevice
     return [MTRBaseDevice deviceWithNodeID:self.nodeID controller:self.deviceController];
 }
 
-// Client Metadata Storage
+#pragma mark - Client Metadata Storage
 
+// REVIEWERS:  does it make sense to constrain this to non-null return and return empty array if necessary?
 - (NSArray *)supportedClientDataClasses
 {
+    // TODO: KMO: check this list, maybe use list from controller DS
+    // TODO: KMO: or, uplevel this out of MTRDevice to another file for use in both places
     return @[ [NSData class], [NSString class], [NSNumber class], [NSDictionary class], [NSArray class] ];
 }
 
+#pragma mark Client Metadata Storage: By Node ID
+// REVIEWERS:  does it make sense to constrain this to non-null return and return empty array if necessary?
+// REVIEWERS:  does it make sense to constrain this to NSArray<NSString *> * since it's keys? kmo 19 jul 2024 09h28
+// TODO:  should this include all keys for node and endpoints on that node?
```

Comment:
@woody-apple regarding this TODO:  the docs for the `clientDataKeys` method and the endpoint-specific keys method are the same in the header, but there was a note in the endpoint-specific keys stub implementation to "enumerate this better" when returning `allKeys` - that sounds like the intention was to return just the nodeID-indexed keys here, and return the nodeID-endpointID indexed keys from the endpoint-specific method.  is that accurate?  if so, can I update the comments for these methods in the header?

Reply:
What's here looks good to me